### PR TITLE
Improve detection of msvc environment when compiling from msvc

### LINKED ISSUE #1308: Compiling from `msbuild` incorrectly uses target build tools for build scripts
The problem arises when running `cargo build` from within a VS project and targeting something other than the native arch (e.g., using `msbuild.exe myproj.sln /p:Platform=otherarch`). When cargo builds a build script, it uses the target linker (added to `PATH` by Visual Studio) instead of the host linker, which naturally fails.

This is a problem as of Rust 1.83 due to the following fix: https://github.com/rust-lang/cc-rs/pull/1201. That PR also contains some discussion about this issue.

### PR DESCRIPTION
**Pull request by dlon**
Fix #1308. This makes path/env detection a bit pickier by not using `PATH` if the compiler has an unexpected target. Previously this caused issues for build scripts since rustc always needs to target the native arch.

### CODE REVIEW

**Comment by ChrisDenton** [comment]
Hmm, I did have to consider this a bit more. On balance, I think I'd rather keep the non-msbuild case the same as it was to avoid unintended side-effects.

So ideally this PR would only affect MSBuild.

**Review by ChrisDenton — COMMENTED** [review]
Ok, I've been testing this and I'm happy that it solves the issue. I do have one small suggestion.

File: src/windows/find_tools.rs [hunk 1]
```diff
@@ -353,20 +353,69 @@ mod impl_ {
...
+        cl.stderr(std::process::Stdio::piped())
+            .stdout(std::process::Stdio::null());
+
+        let out = cl.output().ok()?;
+        let cl_arch = out
+            .stderr
+            .split(|&b| b == b'\n' || b == b'\r')
+            .next()?
+            .split(|&b| b == b' ')
+            .last()?;
```
Comment (ChrisDenton) [inline src/windows/find_tools.rs:410]:
```suggestion
            .rsplit(|&b| b == b' ')
            .next()?
```

Comment (dlon) [reply src/windows/find_tools.rs]: Makes a lot more sense, thanks!

**Review by NobodyXu — APPROVED** [review]
Thanks!

**Pull request merged**